pkg/gadgets/internal/socketenricher/tracer_test.go

```diff
@@ -0,0 +1,249 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package socketenricher
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"testing"
+
+	containerutils "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils"
+	utilstest "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/test"
+	"golang.org/x/sys/unix"
+)
+
+func TestSocketEnricherCreate(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	tracer, err := NewSocketEnricher()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+}
+
+func TestSocketEnricherStopIdempotent(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	tracer, _ := NewSocketEnricher()
+
+	// Check that a double stop doesn't cause issues
+	tracer.Close()
+	tracer.Close()
+}
+
+type sockOpt struct {
+	level int
+	opt   int
+	value int
+}
+
+func TestBindTracer(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	type socketEnricherMapEntry struct {
+		Key   socketenricherSocketsKey
+		Value socketenricherSocketsValue
+	}
+
+	type testDefinition struct {
+		generateEvent func() (uint16, error)
+		validateEvent func(t *testing.T, info *utilstest.RunnerInfo, port uint16, entries []socketEnricherMapEntry)
+	}
+
+	netns, err := containerutils.GetNetNs(os.Getpid())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	stringToSlice := func(s string) (ret [16]int8) {
+		for i := 0; i < 16; i++ {
+			if i >= len(s) {
+				break
+			}
+			ret[i] = int8(s[i])
+		}
+		return
+	}
+
+	for name, test := range map[string]testDefinition{
+		"udp": {
+			generateEvent: bindSocketFn("127.0.0.1", unix.AF_INET, unix.SOCK_DGRAM, 0),
+			validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, port uint16) *socketEnricherMapEntry {
+				return &socketEnricherMapEntry{
+					Key: socketenricherSocketsKey{
+						Netns:  uint32(netns),
+						Family: unix.AF_INET,
+						Proto:  unix.IPPROTO_UDP,
+						Port:   port,
+					},
+					Value: socketenricherSocketsValue{
+						Mntns:   info.MountNsID,
+						PidTgid: uint64(uint32(info.Pid))<<32 + uint64(info.Tid),
+						Task:    stringToSlice("socketenricher."),
+					},
+				}
+			}),
+		},
+		"tcp": {
+			generateEvent: bindSocketFn("127.0.0.1", unix.AF_INET, unix.SOCK_STREAM, 0),
+			validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, port uint16) *socketEnricherMapEntry {
+				return &socketEnricherMapEntry{
+					Key: socketenricherSocketsKey{
+						Netns:  uint32(netns),
+						Family: unix.AF_INET,
+						Proto:  unix.IPPROTO_TCP,
+						Port:   port,
+					},
+					Value: socketenricherSocketsValue{
+						Mntns:   info.MountNsID,
+						PidTgid: uint64(uint32(info.Pid))<<32 + uint64(info.Tid),
+						Task:    stringToSlice("socketenricher."),
+					},
+				}
+			}),
+		},
+		"tcp6": {
```

Comment:
Why not UDP IPv6?

Reply:
It did not work (the bind() syscall returns an error), I don't understand why. I got the tests from `pkg/gadgets/trace/bind/tracer/tracer_test.go` which also does not test UDP IPv6. Maybe the original author (@mauriciovasquezbernal ?) didn't write tests for UDP IPv6 because it also does not work there?

Reply:
It works on my machine :tm: 

```diff
diff --git pkg/gadgets/internal/socketenricher/tracer_test.go pkg/gadgets/internal/socketenricher/tracer_test.go
index 64cebd3f..e3d95a7d 100644
--- pkg/gadgets/internal/socketenricher/tracer_test.go
+++ pkg/gadgets/internal/socketenricher/tracer_test.go
@@ -145,6 +145,24 @@ func TestBindTracer(t *testing.T) {
                                }
                        }),
                },
+               "udp6": {
+                       generateEvent: bindSocketFn("::", unix.AF_INET6, unix.SOCK_DGRAM, 0),
+                       validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, port uint16) *socketEnricherMapEntry {
+                               return &socketEnricherMapEntry{
+                                       Key: socketenricherSocketsKey{
+                                               Netns:  uint32(netns),
+                                               Family: unix.AF_INET6,
+                                               Proto:  unix.IPPROTO_UDP,
+                                               Port:   port,
+                                       },
+                                       Value: socketenricherSocketsValue{
+                                               Mntns:   info.MountNsID,
+                                               PidTgid: uint64(uint32(info.Pid))<<32 + uint64(info.Tid),
+                                               Task:    stringToSlice("socketenricher."),
+                                       },
+                               }
+                       }),
+               },
        } {
                test := test
```

```bash
$ go test -exec sudo . -v -run 'TestBindTracer'
=== RUN   TestBindTracer
=== PAUSE TestBindTracer
=== CONT  TestBindTracer
=== RUN   TestBindTracer/tcp
=== PAUSE TestBindTracer/tcp
=== RUN   TestBindTracer/tcp6
=== PAUSE TestBindTracer/tcp6
=== RUN   TestBindTracer/udp6
=== PAUSE TestBindTracer/udp6
=== RUN   TestBindTracer/udp
=== PAUSE TestBindTracer/udp
=== CONT  TestBindTracer/tcp
=== CONT  TestBindTracer/udp6
=== CONT  TestBindTracer/udp
=== CONT  TestBindTracer/tcp6
--- PASS: TestBindTracer (0.00s)
    --- PASS: TestBindTracer/udp6 (2.92s)
    --- PASS: TestBindTracer/udp (3.01s)
    --- PASS: TestBindTracer/tcp6 (3.03s)
    --- PASS: TestBindTracer/tcp (3.22s)
PASS
ok  	github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher	3.242s
```

Reply:
Thanks @mauriciovasquezbernal ! I will file a separate PR for a udp6 test in pkg/gadgets/trace/bind/tracer/tracer_test.go too.

Reply:
Filed: https://github.com/inspektor-gadget/inspektor-gadget/pull/1290